wxOverlay enhancement

AliKet opened the pull request:
Will replace #2503, #2448 and #2386

Commit:
Make wxOverlay less dependent on external wxDCs

This allows us to use wxDCOverlay like a real dc
and without the need for any pre-constructed wxDC

Commit:
Make wxOverlay (generic impl.) available in all platforms

Will be useful for implementing wxCaret using the generic
overlay rather than the native one for efficiency reasons

Commit:
Use generic wxOverlay to implement wxCaret

This has the benefit of being slightly more efficient
than using native overlays. under Wayland we still need
a native overlay for wxCaret to work

Commit:
caret sample fixup

CreateCaret() should correctly handle caret re-creations

Commit:
Use the new wxDCOverlay's API in drawing sample

Commit:
Use the new wxDCOverlay's API in wxGenericDragImage impl.

with some minor/trivial changes and cleanup

Commit:
Use the new wxDCOverlay's API in wxHeaderCtrl impl.

Commit:
Add native support for wxOverlay under wxMSW

Commit:
Add native support for wxOverlay under wxGTK3

vadz reviewed (commented):
Thanks again for working on this!

I haven't looked at the implementation (including private headers) yet, but the new API looks good to me and if the old API still continues to work (I think it's supposed to, right?), we should try to merge this soon, just documentation is missing.

One thing which is a bit disappointing is that we still need preprocessor checks in `wxDragImage` code, I wonder if why, i.e. what is missing in the generic overlay version?

I'm going to try to finish looking at the code and, ideally, try to test this with my existing code using `wxOverlay` (if you'd like to have a realistic test yourself, please try building https://github.com/vadz/grapheditor/ with this branch and check if it still works).

Thanks again!

AliKet commented on samples/caret/caret.cpp:
```diff
@@ -325,10 +325,11 @@ MyCanvas::~MyCanvas()
     SetCaret(caret);
 
-    caret->Move(m_xMargin, m_yMargin);
+    DoMoveCaret();
```
`MyCanvas::CreateCaret` is also called from `MyFrame::OnSetBlinkTime` which resets the caret's position. and as soon as you start typing again, the caret will jump to the position where it was before you called `OnSetBlinkTime()`. so calling `DoMoveCaret()` (which knows the old position) is what we need here.

AliKet commented on src/generic/dragimgg.cpp:
```diff
@@ -348,6 +335,8 @@ bool wxGenericDragImage::Show()
         // something has changed on the window.
 
 #ifndef wxHAS_NATIVE_OVERLAY
+        wxASSERT_MSG( (m_windowDC != NULL), wxT("No window DC in wxGenericDragImage::Show()") );
```
a copy/past leftover

Commit:
Apply review suggestions

Commit:
Fixups

AliKet commented on include/wx/generic/dragimgg.h:
```diff
@@ -233,12 +233,11 @@ class WXDLLIMPEXP_CORE wxGenericDragImage: public wxObject
     wxWindow*       m_window;
-    wxDC*           m_windowDC;
 
 #ifdef wxHAS_NATIVE_OVERLAY
```
**For one hand**, I've updated `wxGenericDragImage` to use the new (`wxOverlay`) API to see how well (or bad!) it will behave with the changes:
Under **MSW**, `wxGenericDragImage` works perfectly fine (you can hardly distinguish between the native and the the generic version there).

Under **GTK3**, `wxGenericDragImage` works now correctly in fullscreen mode (at least under X11) which was broken before.

**For the other hand**, adapting `wxCaret` to use the generic version of `wxOverlay` was somewhat trivial, which is not the case for `wxGenericDragImage`, and i don't know what to do about it for the moment.

I'll post in the wx-dev about "why wxCaret should use wxOverlay, the generic version and not the native one".

AliKet commented:
@paulcor can you take a look at this again please.

other than the problem i've mentioned [here](https://groups.google.com/g/wx-dev/c/ulLIA9eyBZs),  do you see other problems using this approach to implement wxOverlay under GTK3 ?

TIA.

Commit:
wxGTK3 (X11): Hide the overlay window while moving/resizing

correctly handle the case when the overlay window
is shown and the window it overlays moves or resizes.

Commit:
wxMSW: Hide the overlay window while moving/resizing/iconizing

Commit:
Use wxOverlay for drawing wxGrid selection (optionally)

- also show this in griddemo (Select > Use wxOverlay to draw selection)

- remove the redundant wxVectorGridBlockCoords typedef

- SelectBlockNoEvent() should call SelectBlock() with wxEVT_NULL not false

Commit:
wxCaret: always use native overlay if supported

Now this cancels this previous commit 9dbee97
( Use generic wxOverlay to implement wxCaret )

Commit:
Fixup: handle RTL correctly

Commit:
Fix compilation error

vadz commented:
@AliKet Why is this still in the draft state, do you still intend to change things here or should this be tested/reviewed/merged a.s.a.p.?

paulcor commented:
Doesn't appear to work if you're not using a compositing WM. Overall, this seems too complicated. IMO user code should not need to call things like `IsNative()` or `SetIsManualReset()`.

vadz commented:
I haven't really looked at the latest version yet, but I think the question is whether this is better or worse than e.g. `wxGrid` not working under Wayland at all? Right now I'd settle for any solution that would improve this...

AliKet commented:
> @AliKet Why is this still in the draft state, do you still intend to change things here or should this be tested/reviewed/merged a.s.a.p.?

Yes I have uncommitted changes, and I think I changed my mind and want to undo this commit 89e70c7 (although it works fine)

other than that, testing/reviews are really appreciated, thanks.

AliKet commented:
> Doesn't appear to work if you're not using a compositing WM. 

True but IMHO this shouldn't be a problem since every major WM nowadays comes with integrated compositor by default (as long as the compositor is not explicitly disabled by the used of course).
 and if not, one can still install `xcomposite` extension and use it with the WM.
and I tested `fluxbox + xcomposite` myself and everything works as expected.

> Overall, this seems too complicated. IMO user code should not need to call things like `IsNative()` or `SetIsManualReset()`.

Most of the complication is related to X11 to keep the overlay window (pos & size) in sync with the window it overlays (which is encapsulated in `wxOverlayX11Helper`)
and since this is not needed for something like rubberbanding or dragging, i added `SetIsManualReset()` to enable this class (`wxOverlayX11Helper`) only on demand.

anyhow, `SetIsManualReset()` is removed now and `IsNative()` is for internal use only and shouldn't (normally) be used outside the library.

Thank you.

Commit:
remove SetIsManualReset() API

Commit:
Fixup!

paulcor commented:
> > Doesn't appear to work if you're not using a compositing WM.
> 
> True ...

...

> Most of the complication is related to X11 ...

The thing is, we already have a (generic) overlay implementation that works on X11. How is this complicated new one better?

I'm not at all comfortable with using a toplevel `GdkWindow` as an overlay. It's very likely to be problematic, as they are under the control of the WM. Can you explain why this was necessary?

And this PR adds a ton of code to `wxGrid`, but none of it does what we really need right now, which is to replace drawing on `wxClientDC` with something using `wxOverlay`. In fact, all that code seems to be a way to showcase `wxOverlay`? BTW, it appears to break selection in the `grid` demo: drag-select a few cells, then click another cell, the selection highlight doesn't get removed.

Commit:
fixup! ClearSelection() unintentionally broken when not using overlay

AliKet commented:
> The thing is, we already have a (generic) overlay implementation that works on X11. How is this complicated new one better?
> 

The main problem with the generic implementation is that it doesn't work well with scrolling windows. and I don't know how to fix this unfortunately.

> I'm not at all comfortable with using a toplevel `GdkWindow` as an overlay. It's very likely to be problematic, as they are under the control of the WM. Can you explain why this was necessary?
> 

believe it or not, I spent a lot of time (with no success) trying to come up with a better alternative. at the end I decided to use a toplevel window as overlay just for these advantages:

- Changing `wxWindowGTK` implementation is avoided.
- Isolate the overlay paintings from the main window, i.e. we don't have to use a `wxPaintDC` with the overlay. hence, costly redraws are avoided if all we want is to update the overlay window only.
- No need to save anything to be restored later (as the generic implementation does). and this makes implementing `wxOverlay` to work with scrolled windows easier.
- If the overlay is `Clear()`ed (when rubber-banding or wxGrid-selecting) only the affected region is cleared and/or blitted.
- The system takes care of updating the underlying window when the overlay get destroyed/hidden.
- Under Wayland, creating the overlay as GTK_WINDOW_POPUP will create it as a subsurface window which means the overlay is visually tied to its parent, and is moved/stacked with it.

the problem is only with X11, to sync the overlay's pos & size with the target window.

> And this PR adds a ton of code to `wxGrid`, but none of it does what we really need right now, which is to replace drawing on `wxClientDC` with something using `wxOverlay`. In fact, all that code seems to be a way to showcase `wxOverlay`?

Yes, `wxGrid` is already a complex (beast) piece of code. and modernizing the grid selection using `wxOverlay` was not a trivial task as I initially thought (urgh!). especially I couldn't make the old & the new codes stand under the same umbrella (i.e. share the same code logic) and had to duplicate some functions to get things work correctly, essentially: `GetSelectedBlocksAsRects()` (which looks like `RefreshBlock()`) and its helper `BlockToDeviceRects()` (similar to `BlockToDeviceRect()`).

I'm not going to discuss the wxGrid selection changes here. instead, I'll post a patch against the wxGrid selection stuff using the generic implementation of `wxOverlay` so you can compare/see how hard (if impossible) to go that way (that is, implementing wxGrid selection using the generic overlay)

> BTW, it appears to break selection in the `grid` demo: drag-select a few cells, then click another cell, the selection highlight doesn't get removed.

Fixed now, thanks for noticing this.

Thank you for your patience, time and help to make any progress on this.

AliKet commented:
@paulcor 

Just compare how things work with/without this patch

```
diff --git a/src/common/overlaycmn.cpp b/src/common/overlaycmn.cpp
index 38c364d2bb..717d856973 100644
--- a/src/common/overlaycmn.cpp
+++ b/src/common/overlaycmn.cpp
@@ -217,7 +217,17 @@ void wxDCOverlay::InitFromWindow(wxWindow* win, wxOverlay::Target target, const
         if ( rect.IsEmpty() )
             InitFromDC(dc);
         else
-            InitFromDC(dc, rect.x, rect.y, rect.width, rect.height);
+        {
+            int x = rect.x, y = rect.y;
+            win->ScreenToClient(&x, &y);
+
+            const wxCoord logicalX = dc->DeviceToLogicalX(x);
+            const wxCoord logicalY = dc->DeviceToLogicalY(y);
+            const wxCoord logicalWidth = dc->DeviceToLogicalXRel(rect.width);
+            const wxCoord logicalHeight = dc->DeviceToLogicalYRel(rect.height);
+
+            InitFromDC(dc, logicalX, logicalY, logicalWidth, logicalHeight);
+        }
     }
 }
 
@@ -306,7 +316,7 @@ private:
 void wxOverlay::UseGeneric()
 {
 #ifdef wxHAS_NATIVE_OVERLAY
-    wxASSERT_MSG( !IsOk(), "should only be called for uninitialized overlay" );
+//    wxASSERT_MSG( !IsOk(), "should only be called for uninitialized overlay" );
 
     if ( !m_impl->IsNative() || !m_impl->IsGenericSupported() )
     {
diff --git a/src/generic/grid.cpp b/src/generic/grid.cpp
index ea0edc4030..ea50c03408 100644
--- a/src/generic/grid.cpp
+++ b/src/generic/grid.cpp
@@ -5767,12 +5767,14 @@ void wxGrid::UseOverlay()
     if ( !m_overlay )
     {
         m_overlay = new wxOverlay;
-
+        m_overlay->UseGeneric();
+/*
         if ( !m_overlay->IsNative() )
         {
             delete m_overlay;
             m_overlay = NULL;
         }
+*/
     }
 }
 
```

apply the above patch on top of the current state of the PR
(last commit is 14466ee)

Commit:
Simplify wxOverlayX11Helper implementation

Commit:
Revert (wxCaret) to old implementation
